src/manifest.rs

```diff
@@ -13,7 +14,78 @@ pub(crate) struct Manifest {
 impl Manifest {
     pub(crate) fn load(path: &Path) -> Result<Self, Error> {
         let raw = std::fs::read_to_string(path)?;
-        Ok(toml::from_str(&raw)?)
+        let manifest = toml::from_str(&raw)?;
+        validate_manifest(&manifest)?;
+        Ok(manifest)
+    }
+}
+
+fn validate_manifest(manifest: &Manifest) -> anyhow::Result<()> {
+    use std::fmt::Write;
+
+    let mut name_map: HashMap<&str, Vec<&ManifestFile>> =
+        HashMap::with_capacity(manifest.files.len());
+    let mut url_map: HashMap<&Url, Vec<&ManifestFile>> =
+        HashMap::with_capacity(manifest.files.len());
+    let mut hash_map: HashMap<&str, Vec<&ManifestFile>> =
+        HashMap::with_capacity(manifest.files.len());
+
+    for file in &manifest.files {
+        name_map.entry(&file.name).or_default().push(file);
+        url_map.entry(&file.source).or_default().push(file);
+        hash_map.entry(&file.sha256).or_default().push(file);
+    }
+
+    fn format_conflicts(conflicts: Vec<&ManifestFile>) -> String {
+        conflicts
+            .into_iter()
+            .map(|file| {
+                let ManifestFile {
+                    name,
+                    source,
+                    sha256,
+                } = file;
+                format!(
+                    r#"[[files]]
+name = "{name}"
+source = "{source}"
+sha256 = "{sha256}"
+"#
+                )
+            })
+            .collect::<Vec<_>>()
+            .join("\n")
+    }
+
+    let mut error = String::new();
+    for (name, conflicts) in name_map.into_iter().filter(|(_, files)| files.len() > 1) {
+        writeln!(
+            error,
+            "The following files share the name `{name}`:\n{}\n",
+            format_conflicts(conflicts)
+        )?;
+    }
+    for (url, conflicts) in url_map.into_iter().filter(|(_, files)| files.len() > 1) {
+        writeln!(
+            error,
+            "The following files share the URL `{url}`:\n{}\n",
+            format_conflicts(conflicts)
+        )?;
+    }
+    for (hash, conflicts) in hash_map.into_iter().filter(|(_, files)| files.len() > 1) {
+        writeln!(
+            error,
+            "The following files share the hash `{hash}`:\n{}\n",
+            format_conflicts(conflicts)
+        )?;
+    }
+
+    if error.is_empty() {
+        Ok(())
+    } else {
+        Err(anyhow::anyhow!(
+            "Found error(s) in the file manifest:\n{error}"
+        ))
```

Comment:
Can we instead append the errors to the list of errors in `main.rs`? Ideally I'd love for all errors to appear in a single execution, without an error in the validation step masking errors in the followup steps.